### ISSUE TITLE
Fix failure to generate WMG snapshot for large datasets that throw OOM exception

### ISSUE THREAD

**Issue by prathapsridharan** [chanzuckerberg/single-cell-data-portal#5378]
Here is the stacktrace from the log stream in `staging`. The failure happens in production too and this failurewas replicated in the `staging`:

```
INFO:backend.wmg.pipeline.integrated_corpus.job:Processing dataset 1 of 294
INFO:backend.wmg.pipeline.integrated_corpus.job:h5ad_file_path='datasets/6f7fd0f1-a2ed-4ff1-80d3-33dde731cbc3/local.h5ad'
INFO:backend.wmg.pipeline.integrated_corpus.extract:Extracting datasets/6f7fd0f1-a2ed-4ff1-80d3-33dde731cbc3/local.h5ad...
INFO:backend.wmg.pipeline.integrated_corpus.transform:Applying filters: assay, and lowly-covered cells
ERROR:__main__:Pipeline failed
Traceback (most recent call last):
  File "/backend/wmg/pipeline/cube_pipeline.py", line 88, in main
    corpus_path, stats = load_data_and_create_cube("datasets")
  File "/backend/wmg/pipeline/cube_pipeline.py", line 51, in load_data_and_create_cube
    integrated_corpus.run(path_to_h5ad_datasets, corpus_path, extract_data)
  File "/backend/wmg/pipeline/integrated_corpus/__init__.py", line 15, in run
    build_integrated_corpus(dataset_directory, corpus_path)
  File "/backend/wmg/data/utils.py", line 19, in wrap_func
    result = func(*args, **kwargs)
  File "/backend/wmg/pipeline/integrated_corpus/job.py", line 55, in build_integrated_corpus
    dataset_id, gene_ids = process_h5ad_for_corpus(h5ad_file_path, corpus_path)
  File "/backend/wmg/pipeline/integrated_corpus/job.py", line 98, in process_h5ad_for_corpus
    apply_pre_concatenation_filters(anndata_object)
  File "/backend/wmg/pipeline/integrated_corpus/transform.py", line 31, in apply_pre_concatenation_filters
    scanpy.pp.filter_cells(anndata_object, min_genes=min_genes)
  File "/usr/local/lib/python3.10/site-packages/scanpy/preprocessing/_simple.py", line 133, in filter_cells
    filter_cells(adata.X, min_counts, min_genes, max_counts, max_genes)
  File "/usr/local/lib/python3.10/site-packages/scanpy/preprocessing/_simple.py", line 147, in filter_cells
    X if min_genes is None and max_genes is None else X > 0, axis=1
  File "/usr/local/lib/python3.10/site-packages/scipy/sparse/_compressed.py", line 326, in __gt__
    return self._inequality(other, operator.gt, '_gt_',
  File "/usr/local/lib/python3.10/site-packages/scipy/sparse/_compressed.py", line 297, in _inequality
    return self._scalar_binopt(other, op)
  File "/usr/local/lib/python3.10/site-packages/scipy/sparse/_compressed.py", line 211, in _scalar_binopt
    res = self._with_data(op(self.data, other), copy=True)
  File "/usr/local/lib/python3.10/site-packages/scipy/sparse/_compressed.py", line 1230, in _with_data
    return self.__class__((data, self.indices.copy(),
numpy.core._exceptions._ArrayMemoryError: Unable to allocate 59.5 GiB for an array with shape (7985825460,) and data type int64
```

**Comment by prathapsridharan** [comment]
## Some more info
While attempting to load this dataset in jupyter notebook on EC2 machine that is `c6i.32xlarge` which has about `256 GB` of RAM and running `top -o %MEM` on the machine:

1. While simply reading the anndata object `top` shows `VIRT: 183.3 GB`, `RES: 180 GB` and `%MEM is 72.6%` and `%CPU 100%`. 
2. Next this [line eats up the rest of the memory](https://github.com/scipy/scipy/blob/c025587b9fbbabf7e6bbf10e2e7c76ae61d6395c/scipy/sparse/_compressed.py#L1232) and crashes the process because `copy=True` is [passed here](https://github.com/scipy/scipy/blob/c025587b9fbbabf7e6bbf10e2e7c76ae61d6395c/scipy/sparse/_compressed.py#L213)
3. After patching the problem (2) above, [computing a sum the columns of the matrix](https://github.com/scverse/scanpy/blob/89804c26bcbdde4449dfe98b3193a9d28c2a0e2f/scanpy/preprocessing/_simple.py#L145) by calling `numpy.sum` across `axis=1` eats up the rest of the memory and the process crashes

**Comment by atolopko-czi** [comment]
Can the hosts be configured with swap space?

**Comment by atarashansky** [comment]
I could be reading this wrong but why is [this line](https://github.com/scverse/scanpy/blob/89804c26bcbdde4449dfe98b3193a9d28c2a0e2f/scanpy/preprocessing/_simple.py#L145C25-L145C25) using numpy.sum to sum a SPARSE matrix? This will always take an enormous amount of memory because it needs to materialize the entire array in memory before taking the sum.

That line should read `array.sum(axis=1)`...

**Comment by prathapsridharan** [comment]
@atarashansky - the `np.sum(csr_matrix.data, axis=1)` is what I noted as inefficient after reimplementing `X > 0` such that it doesn't make copies of underlying structures (I didn't really reimplement anything - just passed `copy=False` to a private function that had that option).

Aside from the possibility that my change to `X > 0` can be further improved, I tried replacing `np.sum(csr_matrix.data, axis=1)` with [csr_matrix.sum(axis=1)](https://github.com/scipy/scipy/blob/c025587b9fbbabf7e6bbf10e2e7c76ae61d6395c/scipy/sparse/_compressed.py#L601) hoping that it would more efficient. Unfortunately that doesn't help.

As you mention, it might be the case that both of them make an extra copy when doing summation which doesn't make much sense. Perhaps if the application is multithreaded you could make a case for doing copy on reads but not sure what the reasoning might be if in fact the function is making a copy.

**Comment by atarashansky** [comment]
> I could be reading this wrong but why is [this line](https://github.com/scverse/scanpy/blob/89804c26bcbdde4449dfe98b3193a9d28c2a0e2f/scanpy/preprocessing/_simple.py#L145C25-L145C25) using numpy.sum to sum a SPARSE matrix? This will always take an enormous amount of memory because it needs to materialize the entire array in memory before taking the sum.
> 
> That line should read `array.sum(axis=1)`...
> 
> Also `X > 0` applied to a sparse dataset is also very inefficient.

Based on further investigation and @prathapsridharan 's comment above, np.sum(sparse_matrix) will actually use sparse matrix's internal sum function.

Also, sparse matrix > 0 will return a sparse matrix copy filled with boolean values.

either way, this is crashing us because making a copy of the sparse data causes us to go OOM.

**Comment by prathapsridharan** [comment]
@atarashansky and I worked out another solution to space efficiently filter cells. The general idea is:

1. Use the `anndata.X.indptr`, where `anndata.X` is an instance of `csr_matrix` (sparse row major matrix), to compute the number of non-zero values per row. This requires no extra allocation of memory and is the first major win

2. Remap `anndata.X` to `anndata.raw.X` if `anndata.raw.X` exists because the pipeline exclusively uses `anndata.raw.X` if it is available. Since `size_of(anndata.X) ~= size_of(anndata.raw.X)`, dropping `anndata.X` yields 50% space savings (we reclaim the space by invoking the garbage collector: `gc.collect()`)

**Comment by atolopko-czi** [comment]
Great progress on this! 

From the merged PR, there is a lot of new code w/o test coverage. Is there a plan (or separate issue) to add unit tests? This is touching a critical code path in the pipeline, so just want to make sure we're fully confident we're not introducing any unwanted data loss, changes, etc, on the X data.

**Comment by prathapsridharan** [comment]
@atolopko-czi - There is an epic to improve testing for the wmg pipeline here: https://github.com/chanzuckerberg/single-cell-data-portal/issues/4516. 

Unit tests for these cases are captured here. Presently, there is more urgency to unblock the run. We do have validation tests of the generated snapshot (that is part of the pipeline and run immediately after the data has been generated) to mitigate the risk of exposing bad data to the users.
